optimized_ingestion/stages/detection_estimation/detection_estimation.py

```diff
@@ -109,38 +110,81 @@ def compute_geo_info(self):
+        if exit_time is None or exit_point is None:
+            return None
+        return Action(current_time, exit_time, start_loc=car_loc,
+                      end_loc=exit_point, action_type=CAR_EXIT_SEGMENT,
+                      target_obj_id=self.detection_id,
+                      target_obj_bbox=self.car_bbox2d)
+
+    # def generate_single_sample_action(self, view_distance: float = 50.):
+    #     """Generate a sample plan for the given detection of a single car
+
+    #     Condition 1: detected car is driving towards ego, opposite direction
+    #     Condition 2: detected car driving along the same direction as ego
+    #     Condition 3: detected car and ego are driving into each other
+    #                 i.e. one to north, one from west to it
+    #     Condition 4: detected car and ego are driving away from each other
+    #                 i.e. one to north, one from it to east
+    #     Return: a list of actions
+    #     """
+    #     sample_action_alg = get_sample_action_alg(self.relative_direction)
+    #     if sample_action_alg is not None:
+    #         return self.priority, sample_action_alg(self, view_distance)
+    #     return self.priority, None
 
 
 @dataclass
 class SamplePlan:
     video: "Video"
     next_frame_num: int
     all_detection_info: "List[DetectionInfo]"
+    ego_views: "List[shape.geometry.Polygon]"
+    fps: int = 20
     metadata: Any = None
     current_priority: "float | None" = None
     action: "Action | None" = None
 
+    def update_next_sample_frame_num(self):
+        next_sample_frame_num = len(self.ego_views)
+        assert self.all_detection_info is not None
+        for detection_info in self.all_detection_info:
+            # get the frame num of car exits,
+            # if the detection_info road type is intersection
+            # car_exit_segment_action would be None
+            car_exit_segment_action = detection_info.get_car_exits_segment_action()
+            if car_exit_segment_action and not car_exit_segment_action.invalid_action:
+                car_exit_segment_frame_num = self.find_closest_frame_num(
+                    car_exit_segment_action.finish_time)
+                if car_exit_segment_frame_num > self.next_frame_num:
+                    # get the frame num of car exits view
+                    car_exit_view_frame_num = get_car_exits_view_frame_num(
+                        detection_info, self.ego_views, car_exit_segment_frame_num, self.fps)
+                    next_sample_frame_num = min(next_sample_frame_num,
+                                                car_exit_segment_frame_num,
+                                                car_exit_view_frame_num)
+                else:
+                    next_sample_frame_num = self.next_frame_num
+                    break
+            else:
+                next_sample_frame_num = self.next_frame_num
+                break
+        self.next_frame_num = next_sample_frame_num
+
```

Comment:
This is the major change. Basically, I'm including the entire update next sample frame number logic here. It first finds the frame when the car exits the lane. Then find if the car may exit the camera before that